function/logger.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package function provides an implementation of the telemetry.Logger interface
+// that allows uses a given function to emit logs.
+package function
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/tetratelabs/telemetry"
+)
+
+type (
+	// Emit is a function that will be used to produce log messages by the function Logger.
+	// Implementations of this function just need to implement the log writing. Decisions on whether to
+	// emit a log or not based on the log level should not be made here as the function Logger already
+	// takes care of that.
+	// Similarly, the keyValues parameter presented in this method will already contain al the key/value pairs
+	// that need to be logged.
+	// The function will only be called when the log actually needs to be emitted.
+	Emit func(level telemetry.Level, msg string, err error, values Values)
+
+	// Values contains all the key/value to be included when emitting logs.
+	Values struct {
+		// FromContext has all the key/value pairs that have been added to the Logger Context
+		FromContext []interface{}
+		// FromLogger has all the key/value pairs that have been added to the Logger object itself
+		FromLogger []interface{}
+		// FromMethod has the key/value pairs that were passed to the logging method.
+		FromMethod []interface{}
+	}
+
+	// Logger is an implementation of the telemetry.Logger that allows configuring named
+	// loggers that can be configured independently and referenced by name.
+	Logger struct {
+		// ctx holds the Context to extract key-value pairs from to be added to each
+		// log line.
+		ctx context.Context
+		// args holds the key-value pairs to be added to each log line.
+		args []interface{}
+		// metric holds the Metric to increment each time Info() or Error() is called.
+		metric telemetry.Metric
+		// level holds the configured log level.
+		level *int32
+		// emitFunc is the function that will be used to actually emit the logs
+		emitFunc Emit
+	}
+)
+
+// compile time check for compatibility with the telemetry.Logger interface.
+var _ telemetry.Logger = (*Logger)(nil)
+
+// NewLogger creates a new function Logger that uses the given Emit function to write log messages.
```

Comment:
maybe add to description that a new logger defaults to info level

Reply:
Done